test/orm/test_unitofwork.py

```diff
@@ -3459,3 +3485,21 @@ def test_persistent_flush_sortable(self):
         a.data = "bar"
         b.data = "foo"
         s.commit()
+
+    def test_pep435_custom_sort_key(self):
```

Comment:
I did not implement this test where [you suggested](https://github.com/cans/sqlalchemy/blob/implement-type-engine-value-sortkey/test/sql/test_types.py#L1660) I do. The, maybe wrong, reasons are that:

- the name of the test suite class seemed relevant;
- it simplified the test implementation;
- my understanding is that this happens at the ORM level, I am not sure how to implement this at the core level (though, now that I dug a bit deeper I might have an idea);

Reply:
the test/sql/test_types.py test would be a simple test that you can create an Enum with a "sort_key_function" argument and that you can get it back from the sort_key_function method.   I wonder why I have that as a method.   it should likely be an attribute.

Reply:
Is [this](https://github.com/sqlalchemy/sqlalchemy/pull/4816/files#diff-f02c6a58d783b77b65f72fd0adc49333R1675) what you meant ?

I added a second commit to make an "attribute" (or property) of `sort_key_function`. It's more of a method which default implementation can be `None` (! works because it is used as a callback), than really an attribute, I think. But it does remove a level of indirection and avoids an [unnecessary nested function](https://github.com/sqlalchemy/sqlalchemy/pull/4816/files#diff-c6663807cd4a586af0b59ad165fcfb62R1781). Since the default implementation for the TypeDecorator is to proxy I don't think you can get rid of the `@property` decorator there. We could remove the one in the [TypeEngine](https://github.com/sqlalchemy/sqlalchemy/pull/4816/files#diff-f21ffccc93edfe1d096c18ff1546cedbR374) class though.

Happy to either squash or drop that commit.

Reply:
yeah i guess just a one liner test is what we can do there.

the commits here all get squashed when i pull them into gerrit so you can either squash here or not, I only look at the full product as one patch.

Reply:
Ok. Since I had to make a change to not unnecessarily hit the DB, I squashed the commits.

While running the tests I also hit a tiny issue with `mock`. I managed to fix it with this patch, though I did not include it in this PR as it is off-topic. 
```patch
diff --git a/lib/sqlalchemy/testing/suite/test_types.py b/lib/sqlalchemy/testing/suite/test_types.py
index 82a842102..b8ebcb7ae 100644
--- a/lib/sqlalchemy/testing/suite/test_types.py
+++ b/lib/sqlalchemy/testing/suite/test_types.py
@@ -4,7 +4,7 @@ import datetime
 import decimal
 import json
 
-import mock
+from .. import mock
 
 from .. import config
 from .. import engines
diff --git a/tox.ini b/tox.ini
index ea520f9fe..7eb809bbe 100644
--- a/tox.ini
+++ b/tox.ini
@@ -17,7 +17,7 @@ usedevelop=
 
 deps=pytest!=3.9.1,!=3.9.2
      pytest-xdist
-     mock
+     mock; python_version < '3.3'
      # needed only for correct profiling results
      # due to speed improvements in psycopg2 as of 2.7
      postgresql: psycopg2>=2.7
```
I guess it didn't show before because tox installs mock whatever the python version. But I wanted to be able to just run pytest to get me started so...

Reply:
oh this should be a separate commit I can make and also 1.3 / master is only supporting 3.4 and above, so ill just do that, thanks

Reply:
I guess we need it for 2.7.  sigh